### ISSUE TITLE
[20233] Crash in ROS 2 Windows debug on the 2.11.x branch

### ISSUE THREAD

**Issue by clalancette** [eProsima/Fast-DDS#4230]
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

Windows Debug builds (as in https://ci.ros2.org/view/nightly/job/nightly_win_deb/) should compile and run tests without a crash.

### Current behavior

Windows Debug builds have been crashing since December 20 (https://ci.ros2.org/view/nightly/job/nightly_win_deb/2947/)

### Steps to reproduce

This one is not easy to reproduce, but if you compile all of ROS 2 under Windows Debug, and then run the `test_launch_ros` tests, the crash will happen consistently.

### Fast DDS version/commit

b28171b0223996888f3773d20c51aed94a561d32 (on the 2.11.x branch)

### Platform/Architecture

Windows 10 Visual Studio 2019

### Transport layer

Default configuration, UDPv4 & SHM

### Additional context

We bisected this back to https://github.com/eProsima/Fast-DDS/commit/eae72fd3166406829bc34d14279eea2bf4ab2904 (on the 2.11.x branch).  Prior to that commit, Windows Debug very occasionally crashed, but after that commit it crashes every time.

### Relevant log output

```shell
06:48:45.406 test\test_launch_ros\actions\test_lifecycle_transition.py ..             [  8%]
06:48:45.550 Windows fatal exception: access violation
06:48:45.552 
06:48:45.553 Thread 0x000027c8 (most recent call first):
06:48:45.554   File "C:\Python38\lib\selectors.py", line 314 in _select
06:48:45.555   File "C:\Python38\lib\selectors.py", line 323 in select
06:48:45.557   File "C:\Python38\lib\asyncio\base_events.py", line 1823 in _run_once
06:48:45.558   File "C:\Python38\lib\asyncio\base_events.py", line 570 in run_forever
06:48:45.558   File "C:\ci\ws\install\Lib\site-packages\launch\utilities\signal_management.py", line 138 in run_background_loop
06:48:45.559   File "C:\Python38\lib\threading.py", line 870 in run
06:48:45.560   File "C:\Python38\lib\threading.py", line 932 in _bootstrap_inner
06:48:45.560   File "C:\Python38\lib\threading.py", line 890 in _bootstrap
06:48:45.561 
06:48:45.561 Thread 0x00000950 (most recent call first):
06:48:45.562   File "C:\ci\ws\install\Lib\site-packages\rclpy\executors.py", line 685 in _wait_for_ready_callbacks
06:48:45.563   File "C:\ci\ws\install\Lib\site-packages\rclpy\executors.py", line 781 in wait_for_ready_callbacks
06:48:45.563   File "C:\ci\ws\install\Lib\site-packages\rclpy\executors.py", line 811 in _spin_once_impl
06:48:45.564   File "C:\ci\ws\install\Lib\site-packages\rclpy\executors.py", line 827 in spin_once
06:48:45.564   File "C:\ci\ws\install\Lib\site-packages\rclpy\__init__.py", line 240 in spin
06:48:45.565   File "C:\Python38\lib\threading.py", line 870 in run
06:48:45.565   File "C:\Python38\lib\threading.py", line 932 in _bootstrap_inner
06:48:45.566   File "C:\Python38\lib\threading.py", line 890 in _bootstrap
06:48:45.567 
06:48:45.567 Thread 0x00000740 (most recent call first):
06:48:45.568   File "C:\Python38\lib\concurrent\futures\thread.py", line 78 in _worker
06:48:45.568   File "C:\Python38\lib\threading.py", line 870 in run
06:48:45.569   File "C:\Python38\lib\threading.py", line 932 in _bootstrap_inner
06:48:45.569   File "C:\Python38\lib\threading.py", line 890 in _bootstrap
06:48:45.570 
06:48:45.570 Thread 0x00001fac (most recent call first):
06:48:45.571   File "C:\ci\ws\install\Lib\site-packages\rclpy\node.py", line 187 in __init__
06:48:45.571   File "C:\ci\ws\install\Lib\site-packages\rclpy\__init__.py", line 180 in create_node
06:48:45.572   File "C:\ci\ws\install\Lib\site-packages\launch_ros\ros_adapters.py", line 60 in start
06:48:45.573   File "C:\ci\ws\install\Lib\site-packages\launch_ros\ros_adapters.py", line 52 in __init__
06:48:45.573   File "C:\ci\ws\install\Lib\site-packages\launch_ros\ros_adapters.py", line 119 in get_ros_adapter
06:48:45.574   File "C:\ci\ws\install\Lib\site-packages\launch_ros\ros_adapters.py", line 136 in get_ros_node
06:48:45.575   File "C:\ci\ws\install\Lib\site-packages\launch_ros\actions\load_composable_nodes.py", line 224 in execute
06:48:45.578   File "C:\ci\ws\install\Lib\site-packages\launch\action.py", line 109 in visit
06:48:45.578   File "C:\ci\ws\install\Lib\site-packages\launch\utilities\visit_all_entities_and_collect_futures_impl.py", line 38 in visit_all_entities_and_collect_futures
06:48:45.578   File "C:\ci\ws\install\Lib\site-packages\launch\utilities\visit_all_entities_and_collect_futures_impl.py", line 45 in visit_all_entities_and_collect_futures
06:48:45.578   File "C:\ci\ws\install\Lib\site-packages\launch\utilities\visit_all_entities_and_collect_futures_impl.py", line 45 in visit_all_entities_and_collect_futures
06:48:45.578   File "C:\ci\ws\install\Lib\site-packages\launch\launch_service.py", line 250 in __process_event
06:48:45.578   File "C:\ci\ws\install\Lib\site-packages\launch\launch_service.py", line 230 in _process_one_event
06:48:45.579   File "C:\Python38\lib\asyncio\events.py", line 81 in _run
06:48:45.579   File "C:\Python38\lib\asyncio\base_events.py", line 1851 in _run_once
06:48:45.579   File "C:\Python38\lib\asyncio\base_events.py", line 570 in run_forever
06:48:45.580   File "C:\Python38\lib\asyncio\windows_events.py", line 316 in run_forever
06:48:45.581   File "C:\Python38\lib\asyncio\base_events.py", line 603 in run_until_complete
06:48:45.581   File "C:\ci\ws\install\Lib\site-packages\launch\launch_service.py", line 377 in run
06:48:45.582   File "C:\ci\ws\src\ros2\launch_ros\test_launch_ros\test\test_launch_ros\actions\test_load_composable_nodes.py", line 93 in _assert_launch_no_errors
06:48:45.583   File "C:\ci\ws\src\ros2\launch_ros\test_launch_ros\test\test_launch_ros\actions\test_load_composable_nodes.py", line 132 in test_load_node
06:48:45.583   File "C:\Python38\lib\site-packages\_pytest\python.py", line 194 in pytest_pyfunc_call
06:48:45.583   File "C:\Python38\lib\site-packages\pluggy\_callers.py", line 77 in _multicall
06:48:45.584   File "C:\Python38\lib\site-packages\pluggy\_manager.py", line 115 in _hookexec
06:48:45.584   File "C:\Python38\lib\site-packages\pluggy\_hooks.py", line 493 in __call__
06:48:45.585   File "C:\Python38\lib\site-packages\_pytest\python.py", line 1792 in runtest
06:48:45.585   File "C:\Python38\lib\site-packages\_pytest\runner.py", line 169 in pytest_runtest_call
06:48:45.586   File "C:\Python38\lib\site-packages\pluggy\_callers.py", line 77 in _multicall
06:48:45.586   File "C:\Python38\lib\site-packages\pluggy\_manager.py", line 115 in _hookexec
06:48:45.587   File "C:\Python38\lib\site-packages\pluggy\_hooks.py", line 493 in __call__
06:48:45.587   File "C:\Python38\lib\site-packages\_pytest\runner.py", line 262 in <lambda>
06:48:45.588   File "C:\Python38\lib\site-packages\_pytest\runner.py", line 341 in from_call
06:48:45.588   File "C:\Python38\lib\site-packages\_pytest\runner.py", line 261 in call_runtest_hook
06:48:45.589   File "C:\Python38\lib\site-packages\_pytest\runner.py", line 222 in call_and_report
06:48:45.589   File "C:\Python38\lib\site-packages\_pytest\runner.py", line 133 in runtestprotocol
06:48:45.590   File "C:\Python38\lib\site-packages\pytest_rerunfailures.py", line 541 in pytest_runtest_protocol
06:48:45.590   File "C:\Python38\lib\site-packages\pluggy\_callers.py", line 77 in _multicall
06:48:45.591   File "C:\Python38\lib\site-packages\pluggy\_manager.py", line 115 in _hookexec
06:48:45.591   File "C:\Python38\lib\site-packages\pluggy\_hooks.py", line 493 in __call__
06:48:45.592   File "C:\Python38\lib\site-packages\_pytest\main.py", line 350 in pytest_runtestloop
06:48:45.592   File "C:\Python38\lib\site-packages\pluggy\_callers.py", line 77 in _multicall
06:48:45.593   File "C:\Python38\lib\site-packages\pluggy\_manager.py", line 115 in _hookexec
06:48:45.593   File "C:\Python38\lib\site-packages\pluggy\_hooks.py", line 493 in __call__
06:48:45.594   File "C:\Python38\lib\site-packages\_pytest\main.py", line 325 in _main
06:48:45.595   File "C:\Python38\lib\site-packages\_pytest\main.py", line 271 in wrap_session
06:48:45.595   File "C:\Python38\lib\site-packages\_pytest\main.py", line 318 in pytest_cmdline_main
06:48:45.595   File "C:\Python38\lib\site-packages\pluggy\_callers.py", line 77 in _multicall
06:48:45.596   File "C:\Python38\lib\site-packages\pluggy\_manager.py", line 115 in _hookexec
06:48:45.597   File "C:\Python38\lib\site-packages\pluggy\_hooks.py", line 493 in __call__
06:48:45.597   File "C:\Python38\lib\site-packages\_pytest\config\__init__.py", line 169 in main
06:48:45.597   File "C:\Python38\lib\site-packages\_pytest\config\__init__.py", line 192 in console_main
06:48:45.598   File "C:\Python38\lib\site-packages\pytest\__main__.py", line 5 in <module>
06:48:45.598   File "C:\Python38\lib\runpy.py", line 87 in _run_code
06:48:45.599   File "C:\Python38\lib\runpy.py", line 194 in _run_module_as_main
06:48:46.450 test\test_launch_ros\actions\test_load_composable_nodes.py --- stderr: test_launch_ros
06:48:46.481 Windows fatal exception: access violation
06:48:46.481 
06:48:46.481 Thread 0x000027c8 (most recent call first):
06:48:46.481   File "C:\Python38\lib\selectors.py", line 314 in _select
06:48:46.481   File "C:\Python38\lib\selectors.py", line 323 in select
06:48:46.481   File "C:\Python38\lib\asyncio\base_events.py", line 1823 in _run_once
06:48:46.481   File "C:\Python38\lib\asyncio\base_events.py", line 570 in run_forever
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch\utilities\signal_management.py", line 138 in run_background_loop
06:48:46.481   File "C:\Python38\lib\threading.py", line 870 in run
06:48:46.481   File "C:\Python38\lib\threading.py", line 932 in _bootstrap_inner
06:48:46.481   File "C:\Python38\lib\threading.py", line 890 in _bootstrap
06:48:46.481 
06:48:46.481 Thread 0x00000950 (most recent call first):
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\rclpy\executors.py", line 685 in _wait_for_ready_callbacks
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\rclpy\executors.py", line 781 in wait_for_ready_callbacks
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\rclpy\executors.py", line 811 in _spin_once_impl
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\rclpy\executors.py", line 827 in spin_once
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\rclpy\__init__.py", line 240 in spin
06:48:46.481   File "C:\Python38\lib\threading.py", line 870 in run
06:48:46.481   File "C:\Python38\lib\threading.py", line 932 in _bootstrap_inner
06:48:46.481   File "C:\Python38\lib\threading.py", line 890 in _bootstrap
06:48:46.481 
06:48:46.481 Thread 0x00000740 (most recent call first):
06:48:46.481   File "C:\Python38\lib\concurrent\futures\thread.py", line 78 in _worker
06:48:46.481   File "C:\Python38\lib\threading.py", line 870 in run
06:48:46.481   File "C:\Python38\lib\threading.py", line 932 in _bootstrap_inner
06:48:46.481   File "C:\Python38\lib\threading.py", line 890 in _bootstrap
06:48:46.481 
06:48:46.481 Thread 0x00001fac (most recent call first):
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\rclpy\node.py", line 187 in __init__
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\rclpy\__init__.py", line 180 in create_node
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch_ros\ros_adapters.py", line 60 in start
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch_ros\ros_adapters.py", line 52 in __init__
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch_ros\ros_adapters.py", line 119 in get_ros_adapter
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch_ros\ros_adapters.py", line 136 in get_ros_node
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch_ros\actions\load_composable_nodes.py", line 224 in execute
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch\action.py", line 109 in visit
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch\utilities\visit_all_entities_and_collect_futures_impl.py", line 38 in visit_all_entities_and_collect_futures
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch\utilities\visit_all_entities_and_collect_futures_impl.py", line 45 in visit_all_entities_and_collect_futures
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch\utilities\visit_all_entities_and_collect_futures_impl.py", line 45 in visit_all_entities_and_collect_futures
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch\launch_service.py", line 250 in __process_event
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch\launch_service.py", line 230 in _process_one_event
06:48:46.481   File "C:\Python38\lib\asyncio\events.py", line 81 in _run
06:48:46.481   File "C:\Python38\lib\asyncio\base_events.py", line 1851 in _run_once
06:48:46.481   File "C:\Python38\lib\asyncio\base_events.py", line 570 in run_forever
06:48:46.481   File "C:\Python38\lib\asyncio\windows_events.py", line 316 in run_forever
06:48:46.481   File "C:\Python38\lib\asyncio\base_events.py", line 603 in run_until_complete
06:48:46.481   File "C:\ci\ws\install\Lib\site-packages\launch\launch_service.py", line 377 in run
06:48:46.481   File "C:\ci\ws\src\ros2\launch_ros\test_launch_ros\test\test_launch_ros\actions\test_load_composable_nodes.py", line 93 in _assert_launch_no_errors
06:48:46.481   File "C:\ci\ws\src\ros2\launch_ros\test_launch_ros\test\test_launch_ros\actions\test_load_composable_nodes.py", line 132 in test_load_node
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\python.py", line 194 in pytest_pyfunc_call
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_callers.py", line 77 in _multicall
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_manager.py", line 115 in _hookexec
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_hooks.py", line 493 in __call__
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\python.py", line 1792 in runtest
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\runner.py", line 169 in pytest_runtest_call
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_callers.py", line 77 in _multicall
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_manager.py", line 115 in _hookexec
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_hooks.py", line 493 in __call__
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\runner.py", line 262 in <lambda>
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\runner.py", line 341 in from_call
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\runner.py", line 261 in call_runtest_hook
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\runner.py", line 222 in call_and_report
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\runner.py", line 133 in runtestprotocol
06:48:46.481   File "C:\Python38\lib\site-packages\pytest_rerunfailures.py", line 541 in pytest_runtest_protocol
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_callers.py", line 77 in _multicall
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_manager.py", line 115 in _hookexec
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_hooks.py", line 493 in __call__
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\main.py", line 350 in pytest_runtestloop
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_callers.py", line 77 in _multicall
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_manager.py", line 115 in _hookexec
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_hooks.py", line 493 in __call__
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\main.py", line 325 in _main
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\main.py", line 271 in wrap_session
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\main.py", line 318 in pytest_cmdline_main
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_callers.py", line 77 in _multicall
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_manager.py", line 115 in _hookexec
06:48:46.481   File "C:\Python38\lib\site-packages\pluggy\_hooks.py", line 493 in __call__
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\config\__init__.py", line 169 in main
06:48:46.481   File "C:\Python38\lib\site-packages\_pytest\config\__init__.py", line 192 in console_main
06:48:46.481   File "C:\Python38\lib\site-packages\pytest\__main__.py", line 5 in <module>
06:48:46.481   File "C:\Python38\lib\runpy.py", line 87 in _run_code
06:48:46.481   File "C:\Python38\lib\runpy.py", line 194 in _run_module_as_main
```

**Comment by MiguelCompany** [comment]
@clalancette We already saw this in master, and provided two related fixes on #4182 and #4188.

I just merged the backport for the first one (#4184), and rebased the other one (#4208).

Could you check if building with the branch on #4208 fixes the issue?

**Comment by clalancette** [comment]
> Could you check if building with the branch on #4208 fixes the issue?

Thanks!  Here's a new CI job:

* Windows Debug [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=20790)](http://ci.ros2.org/job/ci_windows/20790/)

**Comment by MiguelCompany** [comment]
> Thanks! Here's a new CI job:
> 
> * Windows Debug [![Build Status](https://camo.githubusercontent.com/320506be7756a6e43dc8875705b0f44fdd3e612f6edc07725b68b735f9a3108c/687474703a2f2f63692e726f73322e6f72672f6275696c645374617475732f69636f6e3f6a6f623d63695f77696e646f7773266275696c643d3230373930)](http://ci.ros2.org/job/ci_windows/20790/)

Did you use branch `mergify/bp/2.11.x/pr-4188` for Fast DDS?

**Comment by clalancette** [comment]
> Did you use branch `mergify/bp/2.11.x/pr-4188` for Fast DDS?

Oh, sorry.  I messed it up and did the wrong thing.  Here is another job:

* Windows Debug [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=20798)](http://ci.ros2.org/job/ci_windows/20798/)

**Comment by MiguelCompany** [comment]
@clalancette

We just merged #4208, even though the last CI you ordered failed. One thing I noticed is the following error message:

`[daemon-stop-1] Failed to load entry point 'stop': No module named 'psutil._psutil_windows'`

So I don't know whether there's also an unrelated infrastructure issue.

On a side note, I opened ros2/ros2#1510 to bump Fast DDS into 2.13.x, since 2.11.x will go EOL soon.

**Comment by clalancette** [comment]
> So I don't know whether there's also an unrelated infrastructure issue.

Oh, it looks like there probably is.  Thanks for pointing it out, I'll take a look.

**Comment by MiguelCompany** [comment]
@clalancette We just merged #4281 which should have fixed this one.

Since it is a backport from a change already present in 2.13.x, I was wondering if we can close this one in favor of ros2/ros2#1510.

But I guess we can just wait for the next run of https://ci.ros2.org/view/nightly/job/nightly_win_deb/

I'll leave the decision to you

**Comment by MiguelCompany** [comment]
I am closing this since the [job](https://ci.ros2.org/view/nightly/job/nightly_win_deb/2985/) after #4281 was merged did not show the crashes originally reported.
